### DISCUSSION TITLE
How to apply control flow in nn.Module attributes?

### DISCUSSION THREAD

**Discussion by wztdream** [google/flax#1613] [Q&A] [answered]
Hi,

It seems flax not allow modify Module attributes in [`setup`](https://flax.readthedocs.io/en/latest/flax.errors.html#flax.errors.SetAttributeFrozenModuleError). As an example, how to wright below code in flax?

```python
class Foo(nn.Module):
    def __init__(self, a=4,b=None)
        self.a = a
        self.b = b
        if b is None:
            self.b = a //2 if a !=0 else 1
```
I tried setting them in `setup` but trigger `SetAttributeInModuleSetupError`
```python
class Foo(nn.Module):
    a:int=4
    b:Any=None
    def setup(self):
        if self.b is None:
            self.b = self.a//2 if self.a!=0 else 1
```

**Comment by avital** [comment, marked as answer]
We should improve the comment on `SetAttributeInModuleSetupError` -- in short, when cloning a module which is necessary for much of the Flax functionality, we need to be able to make a new module instance that has exactly the same attributes. Therefore, if you'd be able to override an attribute and then we'd clone the module, you may get a module instance with different attribute (after it goes through `setup` a second time).

In this case, you'll need to define yet another attribute, like `b_norm` to store the "post-processed" `b` value. Another option is to do the normalization/post-processing within your `__call__` method.

**Reply by wztdream** [reply]
> In this case, you'll need to define yet another attribute, like `b_norm` to store the "post-processed" `b` value. Another option is to do the normalization/post-processing within your `__call__` method.

Thank you, I just found a [commit](https://github.com/google/flax/pull/1206/commits/bc31315cbc57a2e538573d12838fbd5a55c19e50) related with this case, it says we can use user defined `__post_init__` method to post-precoessing the attribute, I tried it and indeed it works. Considering it is possible that one attribute  **depends** on another attribute ( b depends on a in this case), which of the following methods are recommended to **post-processing** attributes? What is the pros and cons between them?

1. I prefer `Foo1` as it modified attributes in  initialization state which is easy to understand. 

2. I am not prefer `Foo2` because add another attributed seems strange to me. 

3. `Foo3` is similar with `Foo2`, we need another variable to hold the post processed value.

```python
class Foo1(nn.Module):
    """ use __post_init__"""
    a: int = 4
    b: Any = None

    def __post_init__(self):
        super().__post_init__()
        if self.b is None:
            self.b = self.a // 2 if self.a != 0 else 1

    @nn.compact
    def __call__(self, x):
        x = nn.Conv(self.b, kernel_size=(3, 3), strides=(1, 1), padding="SAME")(x)
        return x


class Foo2(nn.Module):
   """ use another attribute"""
    a: int = 4
    b: Any = None

    def setup(self):
        if self.b is None:
            self.b_norm = self.a // 2 if self.a != 0 else 1
        else:
            self.b_norm = self.b

    @nn.compact
    def __call__(self, x):
        x = nn.Conv(self.b_norm, kernel_size=(3, 3), strides=(1, 1), padding="SAME")(x)
        return x


class Foo3(nn.Module):
   """post process in __call__"""
    a: int = 4
    b: Any = None

    @nn.compact
    def __call__(self, x):
        if self.b is None:
            b_norm = self.a // 2 if self.a != 0 else 1
        else:
            b_norm = self.b
        x = nn.Conv(b_norm, kernel_size=(3, 3), strides=(1, 1), padding="SAME")(x)
        return x

```

**Reply by avital** [reply]
Hmm I wonder if `Foo1` is actually safe given the need to be able to clone modules. And I'd have to think more carefully, but it's possible that we could relax the limitation of not re-assigning attribute values during setup, which would make your original error simply not exist, by copying the original attributes passed in to a module during construction, and using those in `Module.clone()` rather than the modified attribute values at the time of cloning.

**Reply by wztdream** [reply]
> but it's possible that we could relax the limitation of not re-assigning attribute values during setup, which would make your original error simply not exist

thank you, it will be convenient and easy to understand if the relax is allowed
